logstash/lib/logstash/outputs/loki.rb

```diff
@@ -0,0 +1,262 @@
+  require 'logstash/outputs/loki/entry'
+
+  config_name "loki"
+
+  ## 'Loki URL'
+  config :url, :validate => :string,  :required => true
+
+  ## 'BasicAuth credentials'
+  config :username, :validate => :string, :required => false
+  config :password, :validate => :string, secret: true, :required => false
+
+  ## 'Client certificate'
+  config :cert, :validate => :path, :required => false
+  config :key, :validate => :path, :required => false
+
+  ## 'TLS'
+  config :ca_cert, :validate => :path, :required => false
+
+  ## 'Loki Tenant ID'
+  config :tenant_id, :validate => :string, :default => "fake", :required => false
+
+  ## 'Maximum number of records to hold in a batch before pushing to Loki'
+  config :batch_size, :validate => :number, :default => 10, :required => false
+
+  ## 'Interval in seconds to wait before pushing a batch of records to Loki'
+  config :batch_wait, :validate => :number, :default => 1, :required => false
+
+  ## 'Array of label names to exclude from sending to Loki'
+  config :exclude_labels, :validate => :array, :default => [], :required => false 
+
+  ## 'Extra labels to add to all log streams'
+  config :external_labels, :valide => :hash,  :default => {}, :required => false
+
+
+  public
+  def register
+    @uri = URI.parse(@url + '/loki/api/v1/push')
+    unless @uri.is_a?(URI::HTTP) || @uri.is_a?(URI::HTTPS)
+      raise LogStash::ConfigurationError, "url parameter must be valid HTTP, currently '#{@url}'"
+    end
+
+    @logger.info("New Loki output", :class => self.class.name)
+
+    @Channel = Concurrent::Channel
+    @batches = {}
+    @entries = @Channel.new
+    @exclude_labels += ["message", "@timestamp"] 
+    @retry_http = 0
+    @delay = 1
+    if ssl_cert?
+      load_ssl
+      validate_ssl_key
+    end
+
+    @Channel.go{run()}
+  end
+
+  def ssl_cert?
+    !@key.nil? && !@cert.nil?
+  end
+
+  def load_ssl
+    @cert = OpenSSL::X509::Certificate.new(File.read(@cert)) if @cert
+    @key = OpenSSL::PKey.read(File.read(@key)) if @key
+  end
+
+  def validate_ssl_key
+    if !@key.is_a?(OpenSSL::PKey::RSA) && !@key.is_a?(OpenSSL::PKey::DSA)
+      raise "Unsupported private key type #{key.class}"
+    end
+  end
+
+  def ssl_opts(uri)
+    opts = {
+      use_ssl: uri.scheme == 'https'
+    }
+
+    if !@cert.nil? && !@key.nil?
+      opts = opts.merge(
+        verify_mode: OpenSSL::SSL::VERIFY_PEER,
+        cert: @cert,
+        key: @key
+      )
+    end
+
+    unless @ca_cert.nil?
+      opts = opts.merge(
+        ca_file: @ca_cert
+      )
+    end
+    opts
+  end 
+
+  def run()
+	  min_wait_checkfrequency = 1/1000 #1 millisecond
+	  max_wait_checkfrequency = @batch_wait / 10
+	  if max_wait_checkfrequency < min_wait_checkfrequency
+		  max_wait_checkfrequency = min_wait_checkfrequency
+    end
+
+    max_wait_check = Concurrent::Channel.tick(max_wait_checkfrequency)
+    loop do
+      Concurrent::Channel.select do |s|
+        s.take(@entries) { |e|
```

Comment:
i think you need to add a test covering this.

Tick seems to be expected seconds.

I don’t know if the run function is expected to return to start ? Do you some examples or documentation?

Reply:
Since we are doing this `@Channel.go{run()}` in initialize, it starts a new thread. Same as calling `go func()`. So, main thread returns after this

This is the codebase https://github.com/ruby-concurrency/concurrent-ruby/

You can find some examples here https://github.com/ruby-concurrency/concurrent-ruby/tree/master/examples/go-by-example-channels

Reply:
I was more talking about the ruby plugin documentation for logstash if you have that in hands ;)

Reply:
https://www.elastic.co/guide/en/logstash/current/output-new-plugin.html

Reply:
This is awesome should you set the ` concurrency :single` to make things simpler.